### PR TITLE
Bridge updates for bridgegem system

### LINKED ISSUE #303: Even when specifying OPENC3_TAG=5.1.1 in .env , the latest beta version is built (OpenC3/COSMOS repo)
This is more a cautionary warning than a bug I believe...

I have to use the "OpenC3/cosmos" instead of the reccomended  "OpenC3/cosmos-project" repo, because I want to create the Ethernet To Serial bridge connection, solved and documented in several other posts at BallAerospace/COSMOS.

When cloning that repo, it's the latest beta version (in lmy case 5.1.2-beta) 
Setting OPENC3_TAG=5.1.1 does not prevent building that latest beta version. Is this the desired behavior?

I assume that i must reset the GIT "main" branch to repo to the label "5.1.1" before building OpenC3/cosmos

If this is the desired behavior, I believe it's not a bad Idea to add this to the documentation of Ethernet To Serial connection, 
(all useful issues reside at the BallAerospace/COSMOS repo I 'm afraid)

Greetings, Jurgen

### LINKED ISSUE #354: ERROR MicroserviceOperator: MissingCredentialsError
Hello,

We followed the bridge solution provided in the following link:
>> [Create documentation for bridges (serial ports) · Issue #16 · OpenC3/openc3.com (github.com)](https://github.com/OpenC3/openc3.com/issues/16)

We are currently using the Windows operating system, where we are trying to implement the bridge solution.

We made the following changes to the bridge file:

## cosmos plugin interface:
INTERFACE <%= steam_target_name %>_INT tcpip_client_interface.rb COM21 COM21 10.0 10.0 LENGTH 32 16 17 1 BIG_ENDIAN 

after which we are facing the following issue,

`Unexpected process died... respawning! ruby plugin_microservice.rb
Stdout:
{"time":1672877273126394730,"@timestamp":"2023-01-05T00:07:53.126+00:00","severity":"INFO","microservice_name":"DEFAULT__INTERFACE__STEAM2_INT","container_name":"bc7ac9a8d59d","log":"Microservice initialized with config:
{\"name\"=>\"DEFAULT__INTERFACE__STEAM2_INT\", \"folder_name\"=>nil, \"cmd\"=>[\"ruby\", \"interface_microservice.rb\", \"DEFAULT__INTERFACE__STEAM2_INT\"], \"work_dir\"=>\"/openc3/lib/openc3/microservices\", \"ports\"=>[], \"env\"=>{}, \"topics\"=>[], \"target_names\"=>[\"STEAM2\"], \"options\"=>[], \"container\"=>nil, \"updated_at\"=>1672877261965476398, \"plugin\"=>\"openc3-steam2-1.1.2.20221209010541.gem__20230104011137\", \"needs_dependencies\"=>false}"}

Stderr:
/usr/lib/ruby/gems/3.1.0/gems/aws-sigv4-1.5.2/lib/aws-sigv4/signer.rb:654:in `extract_credentials_provider': missing credentials, provide credentials with one of the following options: (Aws::Sigv4::Errors::MissingCredentialsError)
  - :access_key_id and :secret_access_key
  - :credentials
  - :credentials_provider
    from /usr/lib/ruby/gems/3.1.0/gems/aws-sigv4-1.5.2/lib/aws-sigv4/signer.rb:144:in `initialize'
    from /usr/lib/ruby/gems/3.1.0/gems/aws-sdk-s3-1.114.0/lib/aws-sdk-s3/plugins/s3_signer.rb:235:in `new'
    from /usr/lib/ruby/gems/3.1.0/gems/aws-sdk-s3-1.114.0/lib/aws-sdk-s3/plugins/s3_signer.rb:235:in `build_v4_signer'
    from /usr/lib/ruby/gems/3.1.0/gems/aws-sdk-s3-1.114.0/lib/aws-sdk-s3/plugins/s3_signer.rb:14:in `block in <class:S3Signer>'
    from /usr/lib/ruby/gems/3.1.0/gems/aws-sdk-core-3.158.0/lib/seahorse/client/configuration.rb:72:in `call'
    from /usr/lib/ruby/gems/3.1.0/gems/aws-sdk-core-3.158.0/lib/seahorse/client/configuration.rb:219:in `block in resolve_defaults'
    from /usr/lib/ruby/gems/3.1.0/gems/aws-sdk-core-3.158.0/lib/seahorse/client/configuration.rb:59:in `each'
    from /usr/lib/ruby/gems/3.1.0/gems/aws-sdk-core-3.158.0/lib/seahorse/client/configuration.rb:59:in `each'
    from /usr/lib/ruby/gems/3.1.0/gems/aws-sdk-core-3.158.0/lib/seahorse/client/configuration.rb:218:in `resolve_defaults'
    from /usr/lib/ruby/gems/3.1.0/gems/aws-sdk-core-3.158.0/lib/seahorse/client/configuration.rb:211:in `value_at'
    from /usr/lib/ruby/gems/3.1.0/gems/aws-sdk-core-3.158.0/lib/seahorse/client/configuration.rb:191:in `block in resolve'
    from /usr/lib/ruby/3.1.0/set.rb:511:in `each_key'
    from /usr/lib/ruby/3.1.0/set.rb:511:in `each'
    from /usr/lib/ruby/gems/3.1.0/gems/aws-sdk-core-3.158.0/lib/seahorse/client/configuration.rb:191:in `resolve'
    from /usr/lib/ruby/gems/3.1.0/gems/aws-sdk-core-3.158.0/lib/seahorse/client/configuration.rb:179:in `apply_defaults'
    from /usr/lib/ruby/gems/3.1.0/gems/aws-sdk-core-3.158.0/lib/seahorse/client/configuration.rb:152:in `build!'
    from /usr/lib/ruby/gems/3.1.0/gems/aws-sdk-core-3.158.0/lib/seahorse/client/base.rb:65:in `build_config'
    from /usr/lib/ruby/gems/3.1.0/gems/aws-sdk-core-3.158.0/lib/seahorse/client/base.rb:22:in `initialize'
    from /usr/lib/ruby/gems/3.1.0/gems/aws-sdk-s3-1.114.0/lib/aws-sdk-s3/client.rb:435:in `initialize'
    from /usr/lib/ruby/gems/3.1.0/gems/aws-sdk-core-3.158.0/lib/seahorse/client/base.rb:102:in `new'
    from /usr/lib/ruby/gems/3.1.0/gems/openc3-5.0.11/lib/openc3/microservices/microservice.rb:125:in `initialize'
    from plugin_microservice.rb:27:in `initialize'
    from /usr/lib/ruby/gems/3.1.0/gems/openc3-5.0.11/lib/openc3/microservices/microservice.rb:43:in `new'
    from /usr/lib/ruby/gems/3.1.0/gems/openc3-5.0.11/lib/openc3/microservices/microservice.rb:43:in `run'
    from plugin_microservice.rb:43:in `<main>'`
image

I would highly appreciate your input on the following issue.

### LINKED ISSUE #497: MissingCredentialsError when I try to run bridge
Hi,

I'm running a local version of OpenC3 and when I try to run my bridge I get the following error:

/usr/local/lib/ruby/gems/3.1.0/gems/aws-sigv4-1.5.0/lib/aws-sigv4/signer.rb:646:in `extract_credentials_provider': missing credentials, provide credentials with one of the following options: (Aws::Sigv4::Errors::MissingCredentialsError)
  - :access_key_id and :secret_access_key
  - :credentials
  - :credentials_provider
	from /usr/local/lib/ruby/gems/3.1.0/gems/aws-sigv4-1.5.0/lib/aws-sigv4/signer.rb:136:in `initialize'
	from /usr/local/lib/ruby/gems/3.1.0/gems/aws-sdk-s3-1.114.0/lib/aws-sdk-s3/plugins/s3_signer.rb:235:in `new'
	from /usr/local/lib/ruby/gems/3.1.0/gems/aws-sdk-s3-1.114.0/lib/aws-sdk-s3/plugins/s3_signer.rb:235:in `build_v4_signer'
	from /usr/local/lib/ruby/gems/3.1.0/gems/aws-sdk-s3-1.114.0/lib/aws-sdk-s3/plugins/s3_signer.rb:14:in `block in <class:S3Signer>'
	from /usr/local/lib/ruby/gems/3.1.0/gems/aws-sdk-core-3.131.1/lib/seahorse/client/configuration.rb:72:in `call'
	from /usr/local/lib/ruby/gems/3.1.0/gems/aws-sdk-core-3.131.1/lib/seahorse/client/configuration.rb:219:in `block in resolve_defaults'
	from /usr/local/lib/ruby/gems/3.1.0/gems/aws-sdk-core-3.131.1/lib/seahorse/client/configuration.rb:59:in `each'
	from /usr/local/lib/ruby/gems/3.1.0/gems/aws-sdk-core-3.131.1/lib/seahorse/client/configuration.rb:59:in `each'
	from /usr/local/lib/ruby/gems/3.1.0/gems/aws-sdk-core-3.131.1/lib/seahorse/client/configuration.rb:218:in `resolve_defaults'
	from /usr/local/lib/ruby/gems/3.1.0/gems/aws-sdk-core-3.131.1/lib/seahorse/client/configuration.rb:211:in `value_at'
	from /usr/local/lib/ruby/gems/3.1.0/gems/aws-sdk-core-3.131.1/lib/seahorse/client/configuration.rb:191:in `block in resolve'
	from /usr/local/lib/ruby/3.1.0/set.rb:511:in `each_key'
	from /usr/local/lib/ruby/3.1.0/set.rb:511:in `each'
	from /usr/local/lib/ruby/gems/3.1.0/gems/aws-sdk-core-3.131.1/lib/seahorse/client/configuration.rb:191:in `resolve'
	from /usr/local/lib/ruby/gems/3.1.0/gems/aws-sdk-core-3.131.1/lib/seahorse/client/configuration.rb:179:in `apply_defaults'
	from /usr/local/lib/ruby/gems/3.1.0/gems/aws-sdk-core-3.131.1/lib/seahorse/client/configuration.rb:152:in `build!'
	from /usr/local/lib/ruby/gems/3.1.0/gems/aws-sdk-core-3.131.1/lib/seahorse/client/base.rb:65:in `build_config'
	from /usr/local/lib/ruby/gems/3.1.0/gems/aws-sdk-core-3.131.1/lib/seahorse/client/base.rb:22:in `initialize'
	from /usr/local/lib/ruby/gems/3.1.0/gems/aws-sdk-s3-1.114.0/lib/aws-sdk-s3/client.rb:435:in `initialize'
	from /usr/local/lib/ruby/gems/3.1.0/gems/aws-sdk-core-3.131.1/lib/seahorse/client/base.rb:102:in `new'

I don't understand why I would be getting any AWS error since I'm running local. Not from the Cloud.

Any help would be greatly appreciated!

David

### PR DESCRIPTION
**Pull request by ryanmelt**
closes #497 
closes #354 
closes #303 

Example bridges:
https://github.com/OpenC3/openc3-cosmos-bridge-serial
https://github.com/OpenC3/openc3-cosmos-bridge-hidapi
https://github.com/OpenC3/openc3-cosmos-bridge-dualsense

Bridge documentation: https://openc3.com/docs/v5/bridges

### CODE REVIEW

File: openc3/bin/openc3cli [hunk 1]
```diff
@@ -569,14 +579,32 @@ if not ARGV[0].nil? # argument(s) given
...
+      if s.name == gem_name
+        if Array === Gem.path
+          Gem.path.each do |gem_path|
+            filename = File.join(gem_path, 'gems', "#{s.name}-#{s.version}", 'bridge.txt')
+            puts "Trying #{filename}"
+            break if File.exist?(filename)
+          end
+        else
+          filename = File.join(Gem.path, 'gems', "#{s.name}-#{s.version}", 'bridge.txt')
+        end
```
Comment (jmthomas) [inline openc3/bin/openc3cli:600]: Is this effectively looking for the last gem spec? Is that the latest?

**Review by jmthomas — COMMENTED** [review]
How do we use the new variable system here?

File: openc3/bin/openc3cli (hunk 1 above)
Comment (ryanmelt) [reply openc3/bin/openc3cli:600]: Good catch.  I didn't realize it returns all versions.

Comment (ryanmelt) [reply openc3/bin/openc3cli:600]: It appears to always return the newest version first, but I'll add a comment to that effect.

**Comment by ryanmelt** [comment]
When you run a bridge you can pass in variables like variable1=value1 on the command line

**Comment by ryanmelt** [comment]
Bridge documentation: https://openc3.com/docs/v5/bridges

**Pull request merged**